[pigeon] Use a const for custom type ids for gobject generated files (#156100)

JesseRiemens opened the pull request:
Adds a custom type identifier to generated gobject headers for the user. Calling fl_value_new_custom_object is now possible with that constant.

This fixes https://github.com/flutter/flutter/issues/156100

## Pre-Review Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] page, which explains my responsibilities.
- [X] I read and followed the [relevant style guides] and ran [the auto-formatter].
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I [linked to at least one issue that this PR fixes] in the description above.
- [X] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or I have commented below to indicate which [version change exemption] this PR falls under[^1].
- [X] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or I have commented below to indicate which [CHANGELOG exemption] this PR falls under[^1].
- [X] I updated/added any relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or I have commented below to indicate which [test exemption] this PR falls under[^1].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[^1]: Regular contributors who have demonstrated familiarity with the repository guidelines only need to comment if the PR is not auto-exempted by repo tooling.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[the auto-formatter]: https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[version change exemption]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[CHANGELOG exemption]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test exemption]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

stuartmorgan-g commented on packages/pigeon/platform_tests/test_plugin/linux/pigeon/core_tests.gen.h:
```diff
@@ -1433,6 +1433,23 @@ G_DECLARE_FINAL_TYPE(CoreTestsPigeonTestMessageCodec,
                      CORE_TESTS_PIGEON_TEST, MESSAGE_CODEC,
                      FlStandardMessageCodec)
 
+const uint8_t core_tests_pigeon_test_an_enum_type_id = 129;
```
I haven't done a full review, but from just skimming the generated code changes:
- Why are these `uint8_t` when the ID is an `int` in the `FlValue` headers?
- Constants should be declared (as `extern`) in a header, and defined in the implementation file, rather than have copies inlined into every file including this header.

JesseRiemens commented:
The FLValue headers in some places use `int`, and in some places use `uint8_t`:
```c
// fl_value.h:303
FlValue* fl_value_new_custom(int type, // uses int
                             gconstpointer value,
                             GDestroyNotify destroy_notify);

// fl_standard_message_codec.h:148
gboolean fl_standard_message_codec_write_value(FlStandardMessageCodec* codec,
                                               GByteArray* buffer,  // uses GByteArray*, which is used (subclassed) in generated files:
                                               FlValue* value,
                                               GError** error);
// core_tests.gen.cc:2422
static gboolean
core_tests_pigeon_test_message_codec_write_core_tests_pigeon_test_an_enum(
    FlStandardMessageCodec* codec, GByteArray* buffer, FlValue* value,
    GError** error) {
  uint8_t type = core_tests_pigeon_test_an_enum_type_id; // appends uint8_t to buffer
  g_byte_array_append(buffer, &type, sizeof(uint8_t));
  return fl_standard_message_codec_write_value(codec, buffer, value, error);
}
```
I think this might be related to [#152916](https://github.com/flutter/flutter/issues/152916). Maybe it is best to leave it as is, and for this PR use `int` for the types.

stuartmorgan-g commented:
> The FLValue headers in some places use `int`, and in some places use `uint8_t`:

Where is the header using `uint8_t` for the type? I'm seeing in your snippet:
- A header declaration that uses `int` for the type.
- A header declaration that doesn't have the type at all.
- A file that's not an engine header.

JesseRiemens commented:
My mistake, it's not the FlValue header that uses uint8_t, but only the generated code using the engine header. 
I did not fully understand how it worked, but now I see that the generated dart code at the other side also uses `uint8`:
```dart
          indent.writeln('buffer.putUint8(4);');
```
[dart_generator.dart:432](https://github.com/flutter/packages/blob/5a7d40f877274916bdd81ba18a6b508e2d266aa3/packages/pigeon/lib/src/dart/dart_generator.dart#L432)

That makes more sense to me now. I guess that means we can keep it like it is.

tarrinneal reviewed (commented):
Pr looks good, just one question and I can lgtm

JesseRiemens commented on packages/pigeon/CHANGELOG.md:
```diff
@@ -1,6 +1,7 @@
 
-* Updates minimum supported SDK version to Flutter 3.27/Dart 3.6.
+* [gobject] Adds type id constants in header files so that they can be used by the user.
+* Updates minimum supported SDK version to Flutter 3.24/Dart 3.5.
```
Probably an old version of the tree, I rebased and removed that change.

JesseRiemens commented on packages/pigeon/lib/src/gobject/gobject_generator.dart:
```diff
@@ -338,6 +338,15 @@ class GObjectHeaderGenerator
+
+    final Iterable<EnumeratedType> customTypes =
+        getEnumeratedTypes(root, excludeSealedClasses: true);
+
```
I put a comment, but it's not exactly in gdoc style as there's no direct symbol to refer to. (as the comment concerns a block of consts.) 

If you don't like it, please give me a suggestion and I'll put it in. 

```c
/**
 * Custom type ID constants:
 *
 * Constants used to identify custom types in the codec.
 * They are used in the codec to encode and decode custom types.
 * They may be used in custom object creation functions to identify the type.
 */
```

JesseRiemens commented:
If you'd like to compare, [this commit](https://github.com/flutter/packages/commit/75a101f72801660d911d62036e7eceea5e54bb32) contains the changes since the last review:

stuartmorgan-g reviewed (approved):
LGTM

JesseRiemens commented:
@tarrinneal Can you please autosubmit / merge?
